src/ra_server_meta.erl

```diff
@@ -0,0 +1,211 @@
+%% since this would be prohibitively slow.
+
+-define(FILENAME, "server.meta").
+-define(MAGIC, "RAM1").
+-define(TOTAL_SIZE, 1024).
+-define(LAST_APPLIED_POSITION, ?TOTAL_SIZE - 8).
+-define(TERM_POSITION, ?TOTAL_SIZE - ?LAST_APPLIED_POSITION - 8).
+
+path(DataDir, UId) ->
+    ServerDir = filename:join(DataDir, UId),
+    filename:join(ServerDir, ?FILENAME).
+
+fetch(Path, MetaName, UId) ->
+    case fetch_from_file(Path) of
+        {ok, Metadata} when is_tuple(Metadata) ->
+            {ok, Metadata};
+        {error, _} ->
+            %% metadata migration case:
+            %% fetch from ra_log_meta and store in a file
+            {VotedFor, CurrentTerm, LastApplied} = fetch_from_ra_log_meta(MetaName, UId),
+            case store_sync(Path, VotedFor, CurrentTerm, LastApplied) of
+                ok ->
+                    ra_log_meta:delete(MetaName, UId),
+                    {ok, {VotedFor, CurrentTerm, LastApplied}};
+                Err ->
+                    Err
+            end
+    end.
+
+fetch_from_file(Path) ->
+    case file:read_file(Path) of
+        {ok, <<?MAGIC, VotedForBin:1004/binary, CurrentTerm:64/unsigned, LastApplied:64/unsigned>>} ->
+            VotedFor = try
+                           parse_voted_for(VotedForBin)
+                       catch
+                           _:_ -> undefined
+                       end,
+            {ok, {VotedFor, CurrentTerm, LastApplied}};
+        {ok, _} ->
+            {error, invalid_format};
+        Err ->
+            Err
+    end.
+
+fetch_from_ra_log_meta(MetaName, UId) ->
+    VotedFor = ra_log_meta:fetch(MetaName, UId, voted_for, undefined),
+    CurrentTerm = ra_log_meta:fetch(MetaName, UId, current_term, 0),
+    LastApplied = ra_log_meta:fetch(MetaName, UId, last_applied, 0),
+    {VotedFor, CurrentTerm, LastApplied}.
+
+store_sync(MetaFile, VotedFor, CurrentTerm, LastApplied) when is_binary(MetaFile) ->
+    {ok, MetaFd} = file:open(MetaFile, [write, binary, raw]),
+    store_sync(MetaFd, VotedFor, CurrentTerm, LastApplied),
+    file:close(MetaFd);
+store_sync(MetaFd, VotedFor, CurrentTerm, LastApplied) ->
+    Data = encode_metadata(VotedFor, CurrentTerm, LastApplied),
+    ok = file:pwrite(MetaFd, 0, Data),
+    ok = file:sync(MetaFd).
+
+update_last_applied(MetaFd, LastApplied) ->
+    ok = file:pwrite(MetaFd, ?LAST_APPLIED_POSITION, <<LastApplied:64>>).
+
+encode_metadata(VotedFor, CurrentTerm, LastApplied) ->
+    VotedForBin = case VotedFor of
+                      undefined ->
+                          <<0, 0>>;
+                      {NameAtom, NodeAtom} ->
+                          NameAtomBin = atom_to_binary(NameAtom, utf8),
+                          NodeAtomBin = atom_to_binary(NodeAtom, utf8),
+                          NameSize = byte_size(NameAtomBin),
+                          NodeSize = byte_size(NodeAtomBin),
+                          <<NameSize:8/unsigned, NameAtomBin/binary,
```

Comment:
just noting that although an atom can only contain max 255 (possibly unicode) characters, the binary representation might require more than 255 bytes
```
1> Atom = list_to_atom(lists:duplicate(255, 256)).
'ĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀ'
2> byte_size(atom_to_binary(Atom)).
510
```
term_to_binary stores the atom size in 1 or 2 bytes https://www.erlang.org/doc/apps/erts/erl_ext_dist.html#atom_utf8_ext

sorry if this was considered already, and storing the size in 1 byte is intentional

Reply:
Sorry it's late and I clearly cant math. We'd just need 2052 bytes to encode the atoms so we're still well below the block size.

Reply:
Thanks a lot for pointing this out. I've modified the format and code to use 2 bytes for atom sizes. Technically, if both atoms were full of UTF-8, we could still fail, but what are the odds?! :)  Based on our testing, node names can't be UTF-8 (and the would need to be full of them for this to be a problem)